File: app/src/main/java/biz/logicminds/buelltune/DatabaseVariableProvider.java
/*
 EcmDroid - Android Diagnostic Tool for Buell Motorcycles
 Copyright (C) 2012 by Michel Marti

 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 3
 of the License, or (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, see <http://www.gnu.org/licenses/>.
 */
package biz.logicminds.buelltune;

import android.content.Context;

import biz.logicminds.buelltune.Variable.DataType;
import biz.logicminds.buelltune.data.BitNamesRow;
import biz.logicminds.buelltune.data.EcmDefinitionsDatabase;
import biz.logicminds.buelltune.data.EeVariableRow;
import biz.logicminds.buelltune.data.RtVariableRow;

import java.util.Collection;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Locale;
import java.util.regex.Matcher;

/**
 * Create a Variable based on definitions in the built-in database.
 * <p>
 * Backed by Room (R6, KTD3) instead of the retired {@code DBHelper}. The
 * cache-then-DAO shape is unchanged: the {@link HashMap} caches still sit in
 * front of every lookup because these run on every poll cycle.
 */
public class DatabaseVariableProvider extends VariableProvider {

	private final EcmDefinitionsDatabase db;
	private HashMap<String, Variable> cache = new HashMap<String, Variable>();
	private String current_ecm = null;

	public DatabaseVariableProvider(Context ctx) {
		db = EcmDefinitionsDatabase.getInstance(ctx);
	}

	@Override
	public Collection<String> getRtVariableNames(String ecm) {
		return getRtVariableNames(ecm, null);
	}

	@Override
	public Collection<String> getScalarRtVariableNames(String ecm) {
		return getRtVariableNames(ecm, DataType.SCALAR);
	}

	@Override
	public Collection<String> getBitfieldRtVariableNames(String ecm) {
		return getRtVariableNames(ecm, DataType.BITFIELD);
	}

	private Collection<String> getRtVariableNames(String ecm, DataType type) {
		String typeUpper = type == null ? null : type.toString().toUpperCase(Locale.ENGLISH);
		return new LinkedList<String>(db.rtoffsetsDao().getRtVariableNames(ecm, typeUpper));
	}

	@Override
	public Variable getRtVariable(String ecm, String name) {
		if (ecm != null && !ecm.equals(current_ecm)) {
			cache.clear();
			current_ecm = ecm;
		}
		String key = "rt#" + name;
		Variable ret = cache.get(key);
		if (cache.containsKey(key)) {
			return ret;
		}
		ret = convert(db.rtoffsetsDao().getRtVariable(ecm, name));
		cache.put(key, ret);
		return ret;
	}

	@Override
	public Variable getEEPROMVariable(String ecm, String name) {
		if (ecm == null || name == null) {
			return null;
		}
		if (!ecm.equals(current_ecm)) {
			cache.clear();
			current_ecm = ecm;
		}
		String key = "ee#" + name;
		Variable ret = cache.get(key);
		if (cache.containsKey(key)) {
			return ret;
		}
		ret = convert(db.eeoffsetsDao().getEeVariable(ecm, name));
		cache.put(key, ret);
		return ret;
	}

	@Override
	public Variable getNearestEEPROMVariable(String ecm, int offset) {
		if (ecm == null) {
			return null;
		}
		return convert(db.eeoffsetsDao().getNearestEeVariable(ecm, offset));
	}

	@Override
	public String getName(String varname) {
		Matcher matcher = Constants.BIT_PATTERN.matcher(varname);
		if (matcher.matches()) {
			String name = matcher.group(1);
			int bit = Integer.parseInt(matcher.group(2).split(",")[0]);
			return getName(name, bit);
		}
		return db.namesDao().getName(varname);
	}

	@Override
	public String getName(String varname, int bit) {
		if (bit < 0 || bit > 7) {
			return null;
		}
		BitNamesRow row = db.bitsDao().getBitNames(varname);
		if (row == null) {
			return null;
		}
		switch (bit) {
			case 0: return row.getBitname1();
			case 1: return row.getBitname2();
			case 2: return row.getBitname3();
			case 3: return row.getBitname4();
			case 4: return row.getBitname5();
			case 5: return row.getBitname6();
			case 6: return row.getBitname7();
			default: return row.getBitname8();
		}
	}

	/**
	 * SQLite's TEXT-affinity-to-REAL coercion, applied to the ten
	 * numeric-looking {@code varchar} columns (KTD3.2): a null or
	 * unparseable string reads as {@code 0.0}, matching {@code
	 * Cursor.getDouble()}'s behavior on the same columns before the Room
	 * port.
	 */
	private static double parseDouble(String value) {
		if (value == null) {
			return 0.0;
		}
		try {
			return Double.parseDouble(value);
		} catch (NumberFormatException e) {
			return 0.0;
		}
	}

	/** Same coercion as {@link #parseDouble(String)}, truncated to int - matches {@code Cursor.getInt()} on a TEXT column. */
	private static int parseInt(String value) {
		return (int) parseDouble(value);
	}

	private Variable convert(RtVariableRow row) {
		if (row == null) {
			return null;
		}
		Variable ret = new Variable();
		ret.setId(row.getUniqueid());
		ret.setEcmType(ECM.Type.getType(row.getEcmType()));
		ret.setName(row.getOrigname());
		if (ret.getName() == null) {
			ret.setName(row.getVarname());
		}
		ret.setType(DataType.valueOf(row.getType().toUpperCase(Locale.ENGLISH)));
		ret.setSize(row.getSize());
		ret.setWidth(ret.getSize());
		ret.setOffset(row.getOffset());
		ret.setScale(parseDouble(row.getScale()));
		ret.setTranslate(parseDouble(row.getTranslate()));
		ret.setFormat(row.getFormat());
		ret.setLabel(row.getName());
		ret.setRemarks(row.getRemark());
		ret.setDescription(row.getDescription());
		ret.setUnit(row.getUnits());
		ret.setSymbol(Units.getSymbol(ret.getUnit()));
		ret.setLow(parseDouble(row.getLow()));
		ret.setHigh(parseDouble(row.getHigh()));
		ret.setUlow(parseInt(row.getUlow()));
		ret.setUhigh(parseInt(row.getUhigh()));
		ret.init();
		return ret;
	}

	private Variable convert(EeVariableRow row) {
		if (row == null) {
			return null;
		}
		Variable ret = new Variable();
		ret.setId(row.getUniqueid());
		ret.setEcmType(ECM.Type.getType(row.getEcmType()));
		ret.setName(row.getOrigname());
		if (ret.getName() == null) {
			ret.setName(row.getVarname());
		}
		ret.setType(DataType.valueOf(row.getType().toUpperCase(Locale.ENGLISH)));
		ret.setSize(row.getSize());
		ret.setWidth(row.getElemsize() == null ? 0 : row.getElemsize());
		ret.setCols(row.getCols() == null ? 0 : row.getCols());
		ret.setRows(row.getRows() == null ? 0 : row.getRows());
		ret.setOffset(row.getOffset());
		ret.setScale(parseDouble(row.getScale()));
		ret.setTranslate(parseDouble(row.getTranslate()));
		ret.setFormat(row.getFormat());
		ret.setLabel(row.getName());
		ret.setRemarks(row.getRemark());
		ret.setDescription(row.getDescription());
		ret.setUnit(row.getUnits());
		ret.setSymbol(Units.getSymbol(ret.getUnit()));
		ret.init();
		return ret;
	}
}

File: app/src/main/java/biz/logicminds/buelltune/DatabaseVariableProvider.kt
/*
 EcmDroid - Android Diagnostic Tool for Buell Motorcycles
 Copyright (C) 2012 by Michel Marti

 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 3
 of the License, or (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, see <http://www.gnu.org/licenses/>.
 */
package biz.logicminds.buelltune

import biz.logicminds.buelltune.Variable.DataType
import biz.logicminds.buelltune.data.BitNamesRow
import biz.logicminds.buelltune.data.EcmDefinitionsDatabase
import biz.logicminds.buelltune.data.EeVariableRow
import biz.logicminds.buelltune.data.RtVariableRow
import java.util.LinkedList
import java.util.Locale

/**
 * Create a Variable based on definitions in the built-in database.
 *
 * Backed by Room (R6, KTD3) via the [EcmDefinitionsDatabase] [AppContainer]
 * wires in - the DB lookup itself is unchanged from before this Kotlin port,
 * only how this class obtains the database changed (constructor-injected by
 * [AppContainer] instead of resolving its own `Context`-keyed singleton).
 * The cache-then-DAO shape is unchanged: the [HashMap] caches still sit in
 * front of every lookup because these run on every poll cycle.
 */
class DatabaseVariableProvider(private val db: EcmDefinitionsDatabase) : VariableProvider() {

    private val cache = HashMap<String, Variable?>()
    private var currentEcm: String? = null

    override fun getRtVariableNames(ecm: String): Collection<String> = getRtVariableNames(ecm, null)

    override fun getScalarRtVariableNames(ecm: String): Collection<String> = getRtVariableNames(ecm, DataType.SCALAR)

    override fun getBitfieldRtVariableNames(ecm: String): Collection<String> = getRtVariableNames(ecm, DataType.BITFIELD)

    private fun getRtVariableNames(ecm: String, type: DataType?): Collection<String> {
        val typeUpper = type?.toString()?.uppercase(Locale.ENGLISH)
        return LinkedList(db.rtoffsetsDao().getRtVariableNames(ecm, typeUpper))
    }

    override fun getRtVariable(ecm: String, name: String): Variable? {
        if (ecm != currentEcm) {
            cache.clear()
            currentEcm = ecm
        }
        val key = "rt#$name"
        if (cache.containsKey(key)) {
            return cache[key]
        }
        val ret = convert(db.rtoffsetsDao().getRtVariable(ecm, name))
        cache[key] = ret
        return ret
    }

    override fun getEEPROMVariable(ecm: String, name: String): Variable? {
        if (ecm != currentEcm) {
            cache.clear()
            currentEcm = ecm
        }
        val key = "ee#$name"
        if (cache.containsKey(key)) {
            return cache[key]
        }
        val ret = convert(db.eeoffsetsDao().getEeVariable(ecm, name))
        cache[key] = ret
        return ret
    }

    override fun getNearestEEPROMVariable(ecm: String, offset: Int): Variable? =
        convert(db.eeoffsetsDao().getNearestEeVariable(ecm, offset))

    override fun getName(varname: String): String? {
        val matcher = Constants.BIT_PATTERN.matcher(varname)
        if (matcher.matches()) {
            val name = matcher.group(1)!!
            val bit = matcher.group(2)!!.split(",")[0].toInt()
            return getName(name, bit)
        }
        return db.namesDao().getName(varname)
    }

    override fun getName(varname: String, bitnumber: Int): String? {
        if (bitnumber < 0 || bitnumber > 7) {
            return null
        }
        val row: BitNamesRow = db.bitsDao().getBitNames(varname) ?: return null
        return when (bitnumber) {
            0 -> row.bitname1
            1 -> row.bitname2
            2 -> row.bitname3
            3 -> row.bitname4
            4 -> row.bitname5
            5 -> row.bitname6
            6 -> row.bitname7
            else -> row.bitname8
        }
    }

    private fun convert(row: RtVariableRow?): Variable? {
        if (row == null) return null
        val ret = Variable()
        ret.id = row.uniqueid
        ret.ecmType = ECM.Type.getType(row.ecmType)
        ret.name = row.origname ?: row.varname
        ret.type = DataType.valueOf(row.type.uppercase(Locale.ENGLISH))
        ret.size = row.size
        ret.width = ret.size
        ret.offset = row.offset
        ret.scale = parseDouble(row.scale)
        ret.translate = parseDouble(row.translate)
        ret.format = row.format
        ret.label = row.name
        ret.remarks = row.remark
        ret.description = row.description
        ret.unit = row.units
        ret.symbol = Units.getSymbol(ret.unit)
        ret.low = parseDouble(row.low)
        ret.setHigh(parseDouble(row.high))
        ret.ulow = parseInt(row.ulow)
        ret.uhigh = parseInt(row.uhigh)
        ret.init()
        return ret
    }

    private fun convert(row: EeVariableRow?): Variable? {
        if (row == null) return null
        val ret = Variable()
        ret.id = row.uniqueid
        ret.ecmType = ECM.Type.getType(row.ecmType)
        ret.name = row.origname ?: row.varname
        ret.type = DataType.valueOf(row.type.uppercase(Locale.ENGLISH))
        ret.size = row.size
        ret.width = row.elemsize ?: 0
        ret.cols = row.cols ?: 0
        ret.rows = row.rows ?: 0
        ret.offset = row.offset
        ret.scale = parseDouble(row.scale)
        ret.translate = parseDouble(row.translate)
        ret.format = row.format
        ret.label = row.name
        ret.remarks = row.remark
        ret.description = row.description
        ret.unit = row.units
        ret.symbol = Units.getSymbol(ret.unit)
        ret.init()
        return ret
    }

    companion object {
        /**
         * SQLite's TEXT-affinity-to-REAL coercion, applied to the ten
         * numeric-looking `varchar` columns (KTD3.2): a null or
         * unparseable string reads as 0.0, matching `Cursor.getDouble()`'s
         * behavior on the same columns before the Room port.
         */
        private fun parseDouble(value: String?): Double {
            if (value == null) return 0.0
            return value.toDoubleOrNull() ?: 0.0
        }

        /** Same coercion as [parseDouble], truncated to int - matches `Cursor.getInt()` on a TEXT column. */
        private fun parseInt(value: String?): Int = parseDouble(value).toInt()
    }
}
